arbnode/mel/extraction/messages_in_batch.go

```diff
@@ -0,0 +1,220 @@
+package melextraction
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbcompress"
+	"github.com/offchainlabs/nitro/arbnode/mel"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbos/l1pricing"
+	"github.com/offchainlabs/nitro/arbstate"
+)
+
+// Extracts a list of arbos messages from the sequencer batch, which looks at the
+// segments contained in a batch and extracts the correct ordering of messages
+// from the segments, possibly reading delayed messages when a delayed message
+// virtual segment in encountered.
+func extractMessagesInBatch(
+	ctx context.Context,
+	melState *mel.State,
+	seqMsg *arbstate.SequencerMessage,
+	delayedMsgDB DelayedMessageDatabase,
+) ([]*arbostypes.MessageWithMetadata, error) {
+	var messages []*arbostypes.MessageWithMetadata
+	params := &arbosExtractionParams{
+		melState:         melState,
+		seqMsg:           seqMsg,
+		delayedMsgDB:     delayedMsgDB,
+		targetSubMessage: 0,
+		segmentNum:       0,
+		submessageNumber: 0,
+		blockNumber:      0,
+		timestamp:        0,
+	}
+	var msg *arbostypes.MessageWithMetadata
+	var err error
+	for {
```

Comment:
this loop is constructed after the previous logic, which was sort of fitting for it's purpose (get to the right message and return it) but here it becomes unnecessaarily complicated.
You need as params: timestamp, blockNumber, don't even need to be in a struct
Then loop over segments one by one.
Then according to segment type:
* if empty - skip
* if AdvanceTimestamp or AdvanceBlockNumber - advance that param
* if L2Message | Compressed - read and produce it
* if delayedMessage - read the next one

keeping things like segmentNum, and subMessageNumber etc were useful in the old setup, but if you want a function that just extract all messages from a batch you don't really need them.